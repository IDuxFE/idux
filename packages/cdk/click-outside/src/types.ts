/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type {
  ExtractInnerPropTypes,
  ExtractPublicPropTypes,
  MaybeArray,
  MaybeElement,
  MaybeElementRef,
} from '@idux/cdk/utils'
import type { Component, DefineComponent, HTMLAttributes, ObjectDirective, PropType } from 'vue'

export const clickOutsideProps = {
  disabled: { type: Boolean, default: false },
  is: { type: [String, Object] as PropType<string | Component>, default: 'div' },
  options: { type: Object as PropType<ClickOutsideOptions>, default: undefined },
  onClickOutside: [Function, Array] as PropType<MaybeArray<(evt: MouseEvent) => void>>,
} as const

export type ClickOutsideProps = ExtractInnerPropTypes<typeof clickOutsideProps>
export type ClickOutsidePublicProps = ExtractPublicPropTypes<typeof clickOutsideProps>
export type ClickOutsideComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ClickOutsidePublicProps> & ClickOutsidePublicProps
>
export type ClickOutsideInstance = InstanceType<DefineComponent<ClickOutsideProps>>

export type ClickOutsideDirective = ObjectDirective<MaybeElement, ClickOutsideBinding>

export interface ClickOutsideOptions {
  container?: MaybeElementRef | Window | Document
  exclude?: MaybeElementRef[]
}
export type ClickOutsideHandler = (evt: MouseEvent) => void
export type ClickOutsideBinding = ClickOutsideHandler | (ClickOutsideOptions & { handler: ClickOutsideHandler })
