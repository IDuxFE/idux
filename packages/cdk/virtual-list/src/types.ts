import type { Component, DefineComponent, VNodeTypes } from 'vue'
import type { VueTypeDef } from 'vue-types'

import { PropTypes } from '@idux/cdk/utils'

export interface VirtualItemProps {
  setRef: (element: HTMLElement) => void
}

export const virtualItemPropsDef = {
  setRef: PropTypes.func.def(() => {}),
}

export interface VirtualFillerProps {
  /** Virtual filler height. Should be `count * itemMinHeight` */
  scrollHeight?: number
  /** Set offset of visible items. Should be the top of start item position */
  offset?: number
}

export const virtualFillerPropsDef = {
  scrollHeight: PropTypes.number,
  offset: PropTypes.number,
}

export type VirtualFillerInstance = InstanceType<DefineComponent<VirtualFillerProps>>

export interface VirtualScrollBarProps {
  count: number
  height: number
  scrollHeight: number
  scrollTop: number
}

export const virtualScrollBarPropsDef = {
  count: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  scrollHeight: PropTypes.number.isRequired,
  scrollTop: PropTypes.number.isRequired,
}

export interface VirtualScrollBarBindings {
  delayHidden: () => void
}

export type VirtualScrollBarInstance = InstanceType<DefineComponent<VirtualScrollBarProps, VirtualScrollBarBindings>>

export type ItemRender<T = Record<string, unknown>> = (option: { item: T; index: number }) => VNodeTypes

export interface VirtualListProps<T = Record<string, unknown>> {
  component: string | Component
  data: T[]
  /** If not match virtual scroll condition, Set List use height of container. */
  fullHeight: boolean
  height: number
  itemHeight: number
  itemKey: string | ((item: T) => string | number)
  itemRender?: ItemRender
  /** Set `false` will always use real scroll instead of virtual one */
  virtual: boolean
}

export const virtualListPropsDef = {
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).def('div') as VueTypeDef<string | Component> & {
    required: true
  },
  data: PropTypes.array.def(() => []) as VueTypeDef<Record<string, unknown>[]> & { required: true },
  fullHeight: PropTypes.bool.def(true),
  height: PropTypes.number.def(0),
  itemHeight: PropTypes.number.def(0),
  itemKey: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired as VueTypeDef<
    string | ((item: Record<string, unknown>) => string | number)
  > & { required: true },
  itemRender: PropTypes.func,
  virtual: PropTypes.bool.def(true),
}

export interface VirtualListBindings {
  scrollTo: ScrollToFn
}

export type VirtualListInstance = InstanceType<DefineComponent<VirtualListProps, VirtualListBindings>>

export type ScrollToAlign = 'top' | 'bottom' | 'auto'

export type ScrollToOptions =
  | {
      key: string | number
      align?: ScrollToAlign
      offset?: number
    }
  | {
      index: number
      align?: ScrollToAlign
      offset?: number
    }

export type ScrollToFn = (option?: number | ScrollToOptions) => void
