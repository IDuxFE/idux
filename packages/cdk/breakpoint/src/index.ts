export * from './observe'
export * from './breakpoints'
export * from './mediaMatcher'
