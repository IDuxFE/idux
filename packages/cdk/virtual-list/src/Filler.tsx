import type { CSSProperties } from 'vue'
import type { VirtualFillerProps } from './types'

import { computed, defineComponent, onBeforeUnmount, onMounted, ref } from 'vue'
import { isNil, throttle } from 'lodash-es'
import { offResize, onResize } from '@idux/cdk/utils'
import { virtualFillerProps } from './types'

export default defineComponent({
  name: 'IxVirtualFiller',
  props: virtualFillerProps,
  setup(props, { emit }) {
    const styles = useStyles(props)
    const { innerRef } = useInnerResize(emit)
    return { styles, innerRef }
  },

  render() {
    const { styles, $slots } = this
    return (
      <div class="ix-virtual-filler" style={styles.outer}>
        <div class="ix-virtual-filler-inner" style={styles.inner} ref="innerRef">
          {$slots.default?.()}
        </div>
      </div>
    )
  },
})

const useStyles = (props: VirtualFillerProps) => {
  return computed(() => {
    let outer: CSSProperties = {}
    let inner: CSSProperties = {
      display: 'flex',
      flexDirection: 'column',
    }

    const { offset, scrollHeight = 0 } = props
    if (!isNil(offset) && !isNil(scrollHeight)) {
      outer = { height: `${scrollHeight}px`, position: 'relative', overflow: 'hidden' }

      inner = {
        ...inner,
        transform: `translateY(${offset}px)`,
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
      }
    }
    return { outer, inner }
  })
}

const useInnerResize = (emit: (event: 'resize') => void) => {
  const innerRef = ref<HTMLDivElement>()
  const onInnerResize = throttle(() => emit('resize'), 16)

  onMounted(() => onResize(innerRef.value, onInnerResize))
  onBeforeUnmount(() => offResize(innerRef.value, onInnerResize))

  return { innerRef }
}
