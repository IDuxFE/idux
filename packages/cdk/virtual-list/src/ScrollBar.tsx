import type { ComputedRef, CSSProperties, Ref } from 'vue'
import type { VirtualScrollBarProps } from './types'

import { computed, defineComponent, onBeforeUnmount, onMounted, ref, watch } from 'vue'
import { throttle } from 'lodash-es'
import { on, off, rAF, cancelRAF } from '@idux/cdk/utils'
import { virtualScrollBarProps } from './types'

export default defineComponent({
  name: 'IxVirtualScrollBar',
  props: virtualScrollBarProps,
  setup(props, { emit }) {
    const { visible, delayHidden } = useVisible(props)

    const thumbHight = useThumbHight(props)
    const enableScrollRange = useEnableScrollRange(props)
    const enableHeightRange = useEnableHeightRange(props, thumbHight)
    const thumbTop = useThumbTop(props, enableScrollRange, enableHeightRange)

    const { scrollbarRef, thumbRef, dragging } = useEvents(
      props,
      delayHidden,
      enableScrollRange,
      enableHeightRange,
      thumbTop,
    )

    const scrollbarStyle = useScrollbarStyle(visible)
    const thumbClass = useThumbClass(dragging)
    const thumbStyle = useThumbStyle(thumbHight, thumbTop)

    return { scrollbarRef, scrollbarStyle, thumbRef, thumbClass, thumbStyle, delayHidden }
  },

  render() {
    const { scrollbarStyle, thumbClass, thumbStyle } = this
    return (
      <div ref="scrollbarRef" class="ix-virtual-scrollbar" style={scrollbarStyle}>
        <div ref="thumbRef" class={thumbClass} style={thumbStyle} />
      </div>
    )
  },
})

const getPageY = (evt: MouseEvent | TouchEvent) => {
  return 'touches' in evt ? evt.touches[0].pageY : evt.pageY
}

const useVisible = (props: VirtualScrollBarProps) => {
  const _visible = ref(false)
  let visibleTimeout: NodeJS.Timeout

  const delayHidden = throttle(() => {
    clearTimeout(visibleTimeout)
    _visible.value = true
    visibleTimeout = setTimeout(() => (_visible.value = false), 1200)
  }, 300)

  watch(() => props.scrollTop, delayHidden, { flush: 'post' })

  const visible = computed(() => {
    const { height, scrollHeight } = props
    if (height >= scrollHeight) {
      return false
    }
    return _visible.value
  })

  onBeforeUnmount(() => {
    clearTimeout(visibleTimeout)
  })

  return { visible, delayHidden }
}

const minHight = 20
const useThumbHight = (props: VirtualScrollBarProps) => {
  return computed(() => {
    const { height, count } = props
    let baseHeight = (height / count) * 10
    baseHeight = Math.max(baseHeight, minHight)
    baseHeight = Math.min(baseHeight, height / 2)
    return Math.floor(baseHeight)
  })
}

const useEnableScrollRange = (props: VirtualScrollBarProps) => {
  return computed(() => props.scrollHeight - props.height || 0)
}

const useEnableHeightRange = (props: VirtualScrollBarProps, thumbHight: ComputedRef<number>) => {
  return computed(() => props.height - thumbHight.value || 0)
}

const useThumbTop = (
  props: VirtualScrollBarProps,
  enableScrollRange: ComputedRef<number>,
  enableHeightRange: ComputedRef<number>,
) => {
  return computed(() => {
    const scrollTop = props.scrollTop
    if (scrollTop === 0 || enableScrollRange.value === 0) {
      return 0
    }
    const ptg = scrollTop / enableScrollRange.value
    return ptg * enableHeightRange.value
  })
}

const useEvents = (
  props: VirtualScrollBarProps,
  delayHidden: () => void,
  enableScrollRange: ComputedRef<number>,
  enableHeightRange: ComputedRef<number>,
  thumbTop: ComputedRef<number>,
) => {
  const dragging = ref(false)
  watch(dragging, value => props.onScrollStateChange(value))

  let pageY = 0
  let startTop = 0
  const scrollbarRef = ref<HTMLDivElement>()
  const thumbRef = ref<HTMLDivElement>()
  let rafId: number

  const onScrollbarTouchStart = (evt: TouchEvent) => {
    evt.preventDefault()
  }

  const onScrollbarMouseDown = (evt: MouseEvent) => {
    evt.stopPropagation()
    evt.preventDefault()
  }

  const onScrollbarMouseMove = () => delayHidden()

  const patchEvents = () => {
    on(window, 'mousemove', onMouseMove)
    on(window, 'mouseup', onMouseUp)

    on(thumbRef.value, 'touchmove', onMouseMove)
    on(thumbRef.value, 'touchend', onMouseUp)
  }

  const removeEvents = () => {
    off(window, 'mousemove', onMouseMove)
    off(window, 'mouseup', onMouseUp)

    off(thumbRef.value, 'touchmove', onMouseMove)
    off(thumbRef.value, 'touchend', onMouseUp)

    off(scrollbarRef.value!, 'touchstart', onScrollbarTouchStart)
    off(thumbRef.value, 'touchstart', onMouseDown)

    cancelRAF(rafId)
  }

  const onMouseDown = (evt: MouseEvent | TouchEvent) => {
    dragging.value = true
    pageY = getPageY(evt)
    startTop = thumbTop.value
    patchEvents()
    evt.stopPropagation()
    evt.preventDefault()
  }

  const onMouseMove = (evt: MouseEvent | TouchEvent) => {
    cancelRAF(rafId)
    if (dragging.value) {
      const offsetY = getPageY(evt) - pageY
      const newTop = startTop + offsetY

      const ptg = enableHeightRange.value ? newTop / enableHeightRange.value : 0
      const newScrollTop = Math.ceil(ptg * enableScrollRange.value)

      rafId = rAF(() => props.onScroll(newScrollTop))
    }
  }

  const onMouseUp = () => {
    dragging.value = false
    removeEvents()
  }

  onMounted(() => {
    on(scrollbarRef.value, 'touchstart', onScrollbarTouchStart)
    on(thumbRef.value, 'touchstart', onMouseDown)

    on(scrollbarRef.value, 'mousedown', onScrollbarMouseDown)
    on(scrollbarRef.value, 'mousemove', onScrollbarMouseMove)
    on(thumbRef.value, 'mousedown', onMouseDown)
  })

  onBeforeUnmount(() => {
    removeEvents()

    off(scrollbarRef.value, 'mousedown', onScrollbarMouseDown)
    off(scrollbarRef.value, 'mousemove', onScrollbarMouseMove)
    off(thumbRef.value, 'mousedown', onMouseDown)
  })

  return { scrollbarRef, thumbRef, dragging }
}

const useScrollbarStyle = (visible: ComputedRef<boolean>) => {
  return computed<CSSProperties>(() => {
    return {
      width: '8px',
      top: 0,
      bottom: 0,
      right: 0,
      position: 'absolute',
      display: visible.value ? undefined : 'none',
    }
  })
}

const useThumbClass = (dragging: Ref<boolean>) => {
  return computed(() => {
    return {
      'ix-virtual-scrollbar-thumb': true,
      'ix-virtual-scrollbar-thumb-moving': dragging.value,
    }
  })
}

const useThumbStyle = (thumbHight: ComputedRef<number>, thumbTop: ComputedRef<number>) => {
  return computed<CSSProperties>(() => {
    return {
      width: '100%',
      height: thumbHight.value + 'px',
      top: thumbTop.value + 'px',
      left: 0,
      position: 'absolute',
      background: 'rgba(0, 0, 0, 0.5)',
      borderRadius: '99px',
      cursor: 'pointer',
      userSelect: 'none',
    }
  })
}
