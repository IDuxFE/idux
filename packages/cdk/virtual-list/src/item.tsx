import { defineComponent, cloneVNode } from 'vue'
import { virtualItemPropsDef } from './types'

export default defineComponent({
  name: 'IxVirtualItem',
  props: virtualItemPropsDef,
  render() {
    const { setRef, $slots } = this
    const children = $slots.default?.()
    return children?.length ? cloneVNode(children[0], { ref: setRef }) : null
  },
})
