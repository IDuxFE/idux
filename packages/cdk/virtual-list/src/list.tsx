import type { CSSProperties, ComponentPublicInstance, ComputedRef, Ref } from 'vue'
import type { VirtualItemRenderFn, VirtualFillerInstance, VirtualListProps, VirtualScrollBarInstance } from './types'

import { ref, defineComponent, computed } from 'vue'
import Item from './item'
import Filler from './filler'
import ScrollBar from './scrollBar'
import { virtualListProps } from './types'
import { useEvents } from './utils/useEvents'
import { useItemKey, useItems } from './utils/useItem'
import { useScrollBar } from './utils/useScrollBar'
import { useScrollTo } from './utils/useScrollTo'

export default defineComponent({
  name: 'IxVirtualList',
  props: virtualListProps,
  setup(props) {
    const useVirtual = computed(() => props.virtual && props.height > 0 && props.itemHeight > 0)
    const scrollBarRef = ref<VirtualScrollBarInstance>()
    const { componentRef, componentElement } = useComponentRef()
    const fillerRef = ref<VirtualFillerInstance>()

    const { heights, collectHeight, setItemInstance } = useItems(props)

    const { scrollTop, scrollMoving, scrollState, originScroll, syncScrollTop, setScrollMoving } = useScrollBar(
      props,
      scrollBarRef,
      componentElement,
      fillerRef,
      useVirtual,
      heights,
    )

    const componentStyle = useComponentStyle(props, useVirtual, scrollMoving)

    // When data size reduce. It may trigger native scroll event back to fit scroll position
    const onFallbackScroll = (evt: Event) => {
      const { scrollTop: newScrollTop } = evt.currentTarget as Element
      if (newScrollTop !== scrollTop.value) {
        syncScrollTop(newScrollTop)
      }
      // Trigger origin onScroll
      props.onScroll?.(evt)
    }

    const scrollTo = useScrollTo(props, scrollBarRef, componentElement, heights, collectHeight, syncScrollTop)

    useEvents(componentElement, useVirtual, syncScrollTop, originScroll)

    return {
      useVirtual,
      collectHeight,
      setItemInstance,
      componentRef,
      componentStyle,
      onFallbackScroll,
      fillerRef,
      scrollBarRef,
      scrollState,
      scrollTop,
      syncScrollTop,
      setScrollMoving,
      scrollTo,
    }
  },

  render() {
    const { $props, $slots, data } = this
    const CompTag = this.component as any
    const { start, end, scrollHeight, offset } = this.scrollState
    const itemRender = $slots.item ?? this.itemRender
    const children = useChildren($props, data, start, end, this.setItemInstance, itemRender!)
    return (
      <div class="ix-virtual-list" style={{ position: 'relative' }}>
        <CompTag
          ref="componentRef"
          class="ix-virtual-list-holder"
          style={this.componentStyle}
          onScroll={this.onFallbackScroll}
        >
          <Filler ref="fillerRef" scrollHeight={scrollHeight} offset={offset} onResize={this.collectHeight}>
            {children}
          </Filler>
        </CompTag>
        {this.useVirtual && (
          <ScrollBar
            ref="scrollBarRef"
            count={data.length}
            height={this.height}
            scrollHeight={scrollHeight}
            scrollTop={this.scrollTop}
            onScroll={this.syncScrollTop}
            onScrollStateChange={this.setScrollMoving}
          />
        )}
      </div>
    )
  },
})

const useComponentRef = () => {
  const componentRef = ref<ComponentPublicInstance | HTMLElement>()
  const componentElement = computed<HTMLElement | undefined>(() => {
    const elementOrInstance = componentRef.value
    if (!elementOrInstance) {
      return undefined
    }
    return '$el' in elementOrInstance ? elementOrInstance.$el : elementOrInstance
  })
  return { componentRef, componentElement }
}

const useComponentStyle = (props: VirtualListProps, useVirtual: ComputedRef<boolean>, scrollMoving: Ref<boolean>) => {
  return computed<CSSProperties | null>(() => {
    const { height, fullHeight } = props
    if (height <= 0) {
      return null
    }
    return {
      [fullHeight ? 'height' : 'maxHeight']: height + 'px',
      overflowY: useVirtual.value ? 'hidden' : 'auto',
      overflowAnchor: 'none',
      pointerEvents: useVirtual.value && scrollMoving.value ? 'none' : undefined,
    }
  })
}

const useChildren = (
  props: VirtualListProps,
  data: unknown[],
  startIndex: number,
  endIndex: number,
  setItemRef: (item: unknown, element: HTMLElement) => void,
  itemRender: VirtualItemRenderFn,
) => {
  return data.slice(startIndex, endIndex + 1).map((item, index) => {
    const key = useItemKey(props, item)
    const itemIndex = startIndex + index
    const itemNode = itemRender({ item, index: itemIndex })
    return (
      <Item key={key} setRef={ele => setItemRef(item, ele)}>
        {itemNode}
      </Item>
    )
  })
}
