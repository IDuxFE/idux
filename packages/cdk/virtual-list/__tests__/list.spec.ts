import { h } from 'vue'
import { flushPromises, mount, MountingOptions, VueWrapper } from '@vue/test-utils'
import IxVirtualList from '../src/list'
import { ItemRender, VirtualListInstance, VirtualListProps } from '../src/types'

const getData = (length: number, key = 'id') => {
  const data: { id: string }[] = []
  for (let index = 0; index < length; index++) {
    data.push({ id: `${key}-${index}` })
  }
  return data
}

const defaultProps = {
  height: 200,
  itemHeight: 20,
  itemKey: 'id',
}

const defaultItemSlot = `
<template #item="{ item, index }">
  <span class="virtual-item" style="height: 20px;">{{ item.id }} - {{ index }}</span>
</template>
`

describe('list.ts', () => {
  let VirtualListMount: (options?: MountingOptions<Partial<VirtualListProps>>) => VueWrapper<VirtualListInstance>

  beforeAll(() => {
    jest.spyOn(HTMLElement.prototype, 'clientHeight', 'get').mockImplementation(() => 200)
  })

  afterAll(() => {
    jest.spyOn(HTMLElement.prototype, 'clientHeight', 'get').mockClear()
  })

  beforeEach(() => {
    VirtualListMount = options => mount<VirtualListInstance>(IxVirtualList, { ...options })
  })

  describe('basic work', () => {
    test('render work', () => {
      const wrapper = VirtualListMount({
        props: { data: getData(20), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.html()).toMatchSnapshot()

      expect(() => {
        wrapper.vm.$forceUpdate()
        wrapper.unmount()
      }).not.toThrow()
    })

    test('component work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(20), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.find('.ix-virtual-list-holder').element.tagName).toEqual('DIV')

      await wrapper.setProps({ component: 'ul' })

      expect(wrapper.find('.ix-virtual-list-holder').element.tagName).toEqual('UL')

      await wrapper.setProps({ component: h('span', { class: 'ix-test-span' }) })

      expect(wrapper.find('.ix-virtual-list-holder').element.tagName).toEqual('SPAN')
      expect(wrapper.find('.ix-virtual-list-holder').classes()).toContain('ix-test-span')
    })

    test('data work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(5), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.findAll('.virtual-item').length).toEqual(5)

      await wrapper.setProps({ data: getData(10) })

      expect(wrapper.findAll('.virtual-item').length).toEqual(10)

      await wrapper.setProps({ data: getData(15) })

      expect(wrapper.findAll('.virtual-item').length).toEqual(12)
    })

    test('fullHeight work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(5), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('height: 200px')

      await wrapper.setProps({ fullHeight: false })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('max-height: 200px')

      await wrapper.setProps({ fullHeight: true })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('height: 200px')
    })

    test('height work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(5), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('height: 200px')
      expect(wrapper.findAll('.virtual-item').length).toEqual(5)

      await wrapper.setProps({ data: getData(20) })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('height: 200px')
      expect(wrapper.findAll('.virtual-item').length).toEqual(12)

      await wrapper.setProps({ height: 300 })

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('height: 300px')
      expect(wrapper.findAll('.virtual-item').length).toEqual(17)
    })

    test('itemHeight work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(5), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.findAll('.virtual-item').length).toEqual(5)

      await wrapper.setProps({ data: getData(30) })

      expect(wrapper.findAll('.virtual-item').length).toEqual(12)

      await wrapper.setProps({ itemHeight: 10 })

      expect(wrapper.findAll('.virtual-item').length).toEqual(22)
    })

    test('itemKey work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(20), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      wrapper.findAll('.virtual-item').forEach((item, index) => expect(item.text()).toEqual(`id-${index} - ${index}`))

      await wrapper.setProps({ data: getData(20, 'key'), itemKey: 'key' })

      wrapper.findAll('.virtual-item').forEach((item, index) => expect(item.text()).toEqual(`key-${index} - ${index}`))
    })

    test('itemRender work', async () => {
      const itemRender: ItemRender = ({ item, index }) => {
        return h('span', { class: 'virtual-item' }, [`${item.id} - ${index}`])
      }
      const wrapper = VirtualListMount({
        props: { data: getData(20), ...defaultProps, itemRender },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    test('virtual work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(20), ...defaultProps, virtual: false },
        slots: { item: defaultItemSlot },
      })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeFalsy()

      await wrapper.setProps({ virtual: true })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeTruthy()

      await wrapper.setProps({ itemHeight: 0 })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeFalsy()

      await wrapper.setProps({ itemHeight: 1 })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeTruthy()

      await wrapper.setProps({ height: 0 })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeFalsy()

      await wrapper.setProps({ height: 10 })

      expect(wrapper.find('.ix-virtual-scrollbar').exists()).toBeTruthy()
    })
  })

  describe('scroll work', () => {
    beforeEach(() => {
      jest.useFakeTimers()
    })

    afterEach(() => {
      jest.useRealTimers()
    })

    test('scrollTo work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(40), ...defaultProps },
        slots: { item: defaultItemSlot },
      })
      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(0)
      expect(wrapper.find('.ix-virtual-scrollbar').attributes('style')).toContain('display: none')

      wrapper.vm.scrollTo(100)
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(100)
      expect(wrapper.find('.ix-virtual-scrollbar-thumb').attributes('style')).not.toContain('display: none')

      wrapper.vm.scrollTo({ index: 20, align: 'top' })
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(400)

      wrapper.vm.scrollTo({ index: 20, align: 'bottom' })
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(220)

      wrapper.vm.scrollTo({ key: 'id-20', align: 'top' })
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(400)

      wrapper.vm.scrollTo({ key: 'id-20', align: 'top', offset: 20 })
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(380)

      wrapper.vm.scrollTo(9999)
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(600)

      wrapper.vm.scrollTo(-1)
      jest.runAllTimers()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(0)
    })

    test('moving work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(100), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      wrapper.find('.ix-virtual-scrollbar-thumb').trigger('mousedown', { pageY: 0 })

      const mouseMoveEvent = new Event('mousemove') as MouseEvent
      mouseMoveEvent.pageY = 10
      window.dispatchEvent(mouseMoveEvent)
      await flushPromises()

      expect(wrapper.find('.ix-virtual-list-holder').attributes('style')).toContain('pointer-events: none')

      await flushPromises()
      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop).toEqual(100)

      const mouseUpEvent = new Event('mouseup') as MouseEvent
      window.dispatchEvent(mouseUpEvent)
      await flushPromises()
    })

    test('onScroll work', async () => {
      const onScroll = jest.fn()
      const wrapper = VirtualListMount({
        props: { data: getData(100), ...defaultProps },
        slots: { item: defaultItemSlot },
        attrs: { onScroll },
      })

      wrapper.find('.ix-virtual-list-holder').trigger('scroll')

      expect(onScroll).toBeCalled()
    })
  })

  describe('touch work', () => {
    test('moving work', async () => {
      const wrapper = VirtualListMount({
        props: { data: getData(100), ...defaultProps },
        slots: { item: defaultItemSlot },
      })

      wrapper.find('.ix-virtual-list-holder').trigger('touchstart', { touches: [{ pageY: 100 }] })
      wrapper.find('.ix-virtual-list-holder').trigger('touchmove', { touches: [{ pageY: 80 }] })
      wrapper.find('.ix-virtual-list-holder').trigger('touchend')
      await flushPromises()

      expect(wrapper.find('.ix-virtual-list-holder').element.scrollTop >= 20).toBeTruthy()
    })
  })
})
