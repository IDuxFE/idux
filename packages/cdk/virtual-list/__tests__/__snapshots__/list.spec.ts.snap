// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list.ts basic work itemRender work 1`] = `
"<div class=\\"ix-virtual-list\\" style=\\"position: relative;\\">
  <div class=\\"ix-virtual-list-holder\\" style=\\"height: 200px; overflow-y: hidden;\\">
    <div class=\\"ix-virtual-filler\\" style=\\"height: 400px; position: relative; overflow: hidden;\\">
      <div class=\\"ix-virtual-filler-inner\\" style=\\"display: flex; flex-direction: column; transform: translateY(0px); position: absolute; left: 0px; right: 0px; top: 0px;\\"><span class=\\"virtual-item\\">id-0 - 0</span><span class=\\"virtual-item\\">id-1 - 1</span><span class=\\"virtual-item\\">id-2 - 2</span><span class=\\"virtual-item\\">id-3 - 3</span><span class=\\"virtual-item\\">id-4 - 4</span><span class=\\"virtual-item\\">id-5 - 5</span><span class=\\"virtual-item\\">id-6 - 6</span><span class=\\"virtual-item\\">id-7 - 7</span><span class=\\"virtual-item\\">id-8 - 8</span><span class=\\"virtual-item\\">id-9 - 9</span><span class=\\"virtual-item\\">id-10 - 10</span><span class=\\"virtual-item\\">id-11 - 11</span></div>
    </div>
  </div>
  <div class=\\"ix-virtual-scrollbar\\" style=\\"width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;\\">
    <div class=\\"ix-virtual-scrollbar-thumb\\" style=\\"width: 100%; height: 100px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;\\"></div>
  </div>
</div>"
`;

exports[`list.ts basic work render work 1`] = `
"<div class=\\"ix-virtual-list\\" style=\\"position: relative;\\">
  <div class=\\"ix-virtual-list-holder\\" style=\\"height: 200px; overflow-y: hidden;\\">
    <div class=\\"ix-virtual-filler\\" style=\\"height: 400px; position: relative; overflow: hidden;\\">
      <div class=\\"ix-virtual-filler-inner\\" style=\\"display: flex; flex-direction: column; transform: translateY(0px); position: absolute; left: 0px; right: 0px; top: 0px;\\"><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-0 - 0</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-1 - 1</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-2 - 2</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-3 - 3</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-4 - 4</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-5 - 5</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-6 - 6</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-7 - 7</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-8 - 8</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-9 - 9</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-10 - 10</span><span class=\\"virtual-item\\" style=\\"height: 20px;\\">id-11 - 11</span></div>
    </div>
  </div>
  <div class=\\"ix-virtual-scrollbar\\" style=\\"width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;\\">
    <div class=\\"ix-virtual-scrollbar-thumb\\" style=\\"width: 100%; height: 100px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;\\"></div>
  </div>
</div>"
`;
